synapse/push/bulk_push_rule_evaluator.py

```diff
@@ -342,6 +332,21 @@ async def action_for_event_by_user(
                 # Push rules say we should notify the user of this event
                 actions_by_user[uid] = actions
 
+        # This is a check for the case where user joins a room without being
+        # allowed to see history, and then the server receives a delayed event
+        # from before the user joined, which they should not be pushed for
+        #
+        # We do this *after* calculating the push actions as a) its unlikely
+        # that we'll filter anyone out and b) for large rooms its likely that
+        # most users will have push disabled and so the set of users to check is
+        # much smaller.
+        uids_with_visibility = await filter_event_for_clients_with_state(
+            self.store, actions_by_user.keys(), event, context
+        )
+
+        for user_id in set(uids_with_visibility).difference(uids_with_visibility):
```

Comment:
`set(uids_with_visibility).difference(uids_with_visibility)` is the empty set?

Reply:
AAAAAAAAAAAAAAAAAAAAAAAARGH

Reply:
This requires me to write a test doesn't it

Reply:
I'll write the test after lunch.